src/librustc_mir/build/expr/into.rs

```diff
@@ -156,11 +156,13 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                 //
                 // If `opt_cond_expr` is `None`, then the graph is somewhat simplified:
                 //
-                // [block] --> [loop_block / body_block ] ~~> [body_block_end]    [exit_block]
-                //                         ^                          |
-                //                         |                          |
-                //                         +--------------------------+
+                // [block] --> [loop_block] ~~> [loop_block_end] --false link--> [cleanup_block]
+                //                  ^                   |
+                //                  |                   |
+                //                  +-------------------+
```

Comment:
hmm, this picture seems a bit wrong (I didn't look at the code yet). In particular, if you have something like this:

```rust
loop { continue; }
```

then I think that `loop_block_end` is dead-code, and hence the `FalseEdges` would be stripped out. I think I had expected rather something like:

```rust
[loop_block] --> [body_block] ~~> [loop_block_end]
   |     ^                             |
  false  +------------------------------
   |
   v
  cleanup_block
```

What do you think? (Maybe make a test for this?)

Reply:
When we don't have a condition (i.e. `opt_cond_expr = None`) `loop_block` and `body_block` are one and the same, which wasn't super clear on the original diagram. This code doesn't deal with `continue` or `break` at all, that's handled by some common code in [`scope.rs`](https://github.com/rust-lang/rust/blob/c71cec8834bf30032a8e49d2949f6d8d4080b639/src/librustc_mir/build/scope.rs). You're correct that the loop end block is dead code, but does dead code elimination occur before borrowcheck? If so I can convert the `continue` handling code to also emit a false unwind edge; that should handle the `loop { continue; }` case. Otherwise I don't think we have a problem.

Reply:
> This code doesn't deal with continue or break at all

Well, it sets up the structures that tell that code where to break to. This is what the call to `in_breakable_scope` is doing:

https://github.com/rust-lang/rust/blob/def3269a71be2e737cad27418a3dad9f5bd6cd32/src/librustc_mir/build/expr/into.rs#L172-L173

In particular, the `Some(loop_block)` says "on continue, branch here". 

> You're correct that the loop end block is dead code, but does dead code elimination occur before borrowcheck? 

It's not particularly relevant whether dead code is removed; the goal of inserting these false edges is to ensure that every path from the START block reached either an unwind point or the return point (when false edges are included). 

> If so I can convert the continue handling code to also emit a false unwind edge; that should handle the loop { continue; } case. Otherwise I don't think we have a problem.

You could, but I don't understand why that would be better. It seems like it's a bit more complicated to implement and also results in more blocks/false edges overall. Am I missing something?